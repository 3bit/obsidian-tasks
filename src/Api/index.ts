import type { App } from 'obsidian';
import { createOrEditTaskLineModal } from './createOrEditTaskLineModal';
import type { TasksApiV1 } from './TasksApiV1';

/**
 * Factory method for API v1
 *
 * @param app - The Obsidian App
 */
export const tasksApiV1 = (app: App): TasksApiV1 => {
    return {
        createOrEditTaskLineModal: (taskLine?: string): Promise<string> => {
            return createOrEditTaskLineModal(app, taskLine);
        },
    };
};
