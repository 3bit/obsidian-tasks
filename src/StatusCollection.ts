/**
 * The type used for a single entry in bulk imports of pre-created sets of statuses, such as for Themes or CSS Snippets.
 */
export type StatusCollectionEntry = [string, string, string];

/**
 * The type used for bulk imports of pre-created sets of statuses, such as for Themes or CSS Snippets.
 */
export type StatusCollection = Array<StatusCollectionEntry>;
