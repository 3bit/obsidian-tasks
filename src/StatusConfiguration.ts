/**
 * This is the object stored by the Obsidian configuration and used to create the status
 * objects for the session
 *
 * @export
 * @class StatusConfiguration
 */
export class StatusConfiguration {
    /**
     * The indicator used between the two square brackets in the markdown task.
     *
     * @type {string}
     * @memberof Status
     */
    public readonly indicator: string;

    /**
     * Returns the name of the status for display purposes.
     *
     * @type {string}
     * @memberof Status
     */
    public readonly name: string;

    /**
     * Returns the next status for a task when toggled.
     *
     * @type {string}
     * @memberof Status
     */
    public readonly nextStatusIndicator: string;

    /**
     * If true then it is registered as a command that the user can map to.
     *
     * @type {boolean}
     * @memberof Status
     */
    public readonly availableAsCommand: boolean;

    /**
     * Creates an instance of Status. The registry will be added later in the case
     * of the default statuses.
     *
     * @param {string} indicator
     * @param {string} name
     * @param {Status} nextStatusIndicator
     * @param {boolean} availableAsCommand
     * @memberof Status
     */
    constructor(indicator: string, name: string, nextStatusIndicator: string, availableAsCommand: boolean) {
        this.indicator = indicator;
        this.name = name;
        this.nextStatusIndicator = nextStatusIndicator;
        this.availableAsCommand = availableAsCommand;
    }
}
