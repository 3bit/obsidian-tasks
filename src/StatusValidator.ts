import type { StatusConfiguration } from './StatusConfiguration';

export class StatusValidator {
    /**
     * Determine whether the date in this object is valid, and return error message(s) for display if not.
     */
    public validate(statusConfiguration: StatusConfiguration): string[] {
        const errors: string[] = [];

        // Messages are added in the order fields are shown when editing statuses.
        errors.push(...this.validateIndicator(statusConfiguration));
        errors.push(...this.validateName(statusConfiguration));
        errors.push(...this.validateNextIndicator(statusConfiguration));

        return errors;
    }

    public validateIndicator(statusConfiguration: StatusConfiguration): string[] {
        return StatusValidator.validateOneIndicator(statusConfiguration.indicator, 'Task Status Symbol');
    }

    public validateNextIndicator(statusConfiguration: StatusConfiguration): string[] {
        return StatusValidator.validateOneIndicator(statusConfiguration.nextStatusIndicator, 'Task Next Status Symbol');
    }

    public validateName(statusConfiguration: StatusConfiguration) {
        const errors: string[] = [];
        if (statusConfiguration.name.length === 0) {
            errors.push('Task Status Name cannot be empty.');
        }
        return errors;
    }

    private static validateOneIndicator(indicator: string, indicatorName: string): string[] {
        const errors: string[] = [];
        if (indicator.length === 0) {
            errors.push(`${indicatorName} cannot be empty.`);
        }

        if (indicator.length > 1) {
            errors.push(`${indicatorName} ("${indicator}") must be a single character.`);
        }
        return errors;
    }
}
