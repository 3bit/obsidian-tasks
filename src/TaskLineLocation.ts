/**
 * TaskLineLocation will evolve in to the place where all information about a task line's location
 * in a markdown file will be stored, so that testable algorithms can then be added here.
 */
export class TaskLineLocation {
    private readonly _path: string;

    public constructor(path: string) {
        this._path = path;
    }

    public get path(): string {
        return this._path;
    }
}
