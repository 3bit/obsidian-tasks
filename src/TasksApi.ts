import type { App } from 'obsidian';
import type { Task } from './Task';
import { TaskModal } from './TaskModal';
import { taskFromLine } from './Commands/CreateOrEditTaskParser';

export class TasksApi {
    public static GetApi(app: App) {
        return {
            showUi: (): Promise<string> => {
                return this.showUi(app);
            },
        };
    }

    public static showUi(app: App): Promise<string> {
        let resolvePromise: (input: string) => void;
        const waitForClose = new Promise<string>((resolve, _) => {
            resolvePromise = resolve;
        });

        const onSubmit = (updatedTasks: Task[]): void => {
            const line = updatedTasks.map((task: Task) => task.toFileLineString()).join('\n');
            resolvePromise(line);
        };

        const taskModal = new TaskModal({
            app,
            task: taskFromLine({ line: '', path: '' }),
            onSubmit,
        });
        taskModal.open();
        return waitForClose;
    }
}
