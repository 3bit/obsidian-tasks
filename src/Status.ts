import { StatusConfiguration, StatusType } from './StatusConfiguration';

/**
 * Tracks the possible states that a task can be in.
 *
 * Related classes:
 * @see StatusConfiguration
 * @see StatusRegistry
 * @see StatusSettings
 * @see StatusSettingsHelpers.ts
 * @see CustomStatusModal
 *
 * @export
 * @class Status
 */
export class Status {
    /**
     * The default Done status. Goes to Todo when toggled.
     *
     * @static
     * @type {Status}
     * @memberof Status
     */
    public static DONE: Status = Status.makeDone();

    /**
     * A default status of empty, used when things go wrong.
     *
     * @static
     * @memberof Status
     */
    public static EMPTY: Status = Status.makeEmpty();

    /**
     * The default Todo status. Goes to Done when toggled.
     * User may later be able to override this to go to In Progress instead.
     *
     * @static
     * @type {Status}
     * @memberof Status
     */
    public static TODO: Status = Status.makeTodo();

    /**
     * The configuration stored in the data.json file.
     *
     * @type {StatusConfiguration}
     * @memberof Status
     */
    public readonly configuration: StatusConfiguration;

    /**
     * The indicator used between the two square brackets in the markdown task.
     *
     * @type {string}
     * @memberof Status
     */
    public get indicator(): string {
        return this.configuration.indicator;
    }

    /**
     * Returns the name of the status for display purposes.
     *
     * @type {string}
     * @memberof Status
     */
    public get name(): string {
        return this.configuration.name;
    }

    /**
     * Returns the next status for a task when toggled.
     *
     * @type {string}
     * @memberof Status
     */
    public get nextStatusIndicator(): string {
        return this.configuration.nextStatusIndicator;
    }

    /**
     * If true then it is registered as a command that the user can map to.
     *
     * @type {boolean}
     * @memberof Status
     */
    public get availableAsCommand(): boolean {
        return this.configuration.availableAsCommand;
    }

    /**
     * Returns the status type. See {@link StatusType} for details.
     */
    public get type(): StatusType {
        return this.configuration.type;
    }

    /**
     * Creates an instance of Status. The registry will be added later in the case
     * of the default statuses.
     *
     * @param {StatusConfiguration} configuration
     * @memberof Status
     */
    constructor(configuration: StatusConfiguration) {
        this.configuration = configuration;
    }

    /**
     * The default Done status. Goes to Todo when toggled.
     */
    static makeDone(): Status {
        return new Status(new StatusConfiguration('x', 'Done', ' ', true, StatusType.DONE));
    }

    /**
     * A default status of empty, used when things go wrong.
     */
    static makeEmpty(): Status {
        return new Status(new StatusConfiguration('', 'EMPTY', '', true, StatusType.EMPTY));
    }

    /**
     * The default Todo status. Goes to Done when toggled.
     * User may later be able to override this to go to In Progress instead.
     */
    static makeTodo(): Status {
        return new Status(new StatusConfiguration(' ', 'Todo', 'x', true, StatusType.TODO));
    }

    /**
     * The default Cancelled status. Goes to Todo when toggled.
     */
    static makeCancelled(): Status {
        return new Status(new StatusConfiguration('-', 'Cancelled', ' ', true, StatusType.CANCELLED));
    }

    /**
     * The default In Progress status. Goes to Done when toggled.
     */
    static makeInProgress(): Status {
        return new Status(new StatusConfiguration('/', 'In Progress', 'x', true, StatusType.IN_PROGRESS));
    }

    /**
     * Return the StatusType to use for an indicator, if it is not in the StatusRegistry.
     * The core indicators are recognised.
     * Other indicators are treated as StatusType.TODO
     * @param indicator
     */
    static getTypeForUnknownIndicator(indicator: string): StatusType {
        switch (indicator) {
            case 'x':
            case 'X':
                return StatusType.DONE;
            case '/':
                return StatusType.IN_PROGRESS;
            case '-':
                return StatusType.CANCELLED;
            case '':
                return StatusType.EMPTY;
            case ' ':
            default:
                return StatusType.TODO;
        }
    }

    /**
     * Create a Status representing the given, unknown indicator.
     *
     * This can be useful when StatusRegistry does not recognise an indicator,
     * and we do not want to expose the user's data to the Status.EMPTY status.
     * @param unknownIndicator
     */
    static createUnknownStatus(unknownIndicator: string) {
        return new Status(new StatusConfiguration(unknownIndicator, 'Unknown', 'x', false));
    }

    /**
     * Returns the completion status for a task, this is only supported
     * when the task is done/x.
     *
     * @return {*}  {boolean}
     * @memberof Status
     */
    public isCompleted(): boolean {
        return this.indicator === 'x' || this.indicator === 'X';
    }

    /**
     * Return a one-line summary of the status, for presentation to users.
     */
    public previewText() {
        let commandNotice = '';
        if (Status.tasksPluginCanCreateCommandsForStatuses() && this.availableAsCommand) {
            commandNotice = 'Available as a command.';
        }
        return `- [${this.indicator}] ${this.name}, next status is '${this.nextStatusIndicator}', type is '${this.configuration.type}'. ${commandNotice}`;
    }

    /**
     * Whether Tasks can yet create 'Toggle Status' commands for statuses
     *
     * This is not yet possible, and so some UI features are temporarily hidden.
     * See https://github.com/obsidian-tasks-group/obsidian-tasks/issues/1486
     * Once that issue is addressed, this method can be removed.
     */
    public static tasksPluginCanCreateCommandsForStatuses(): boolean {
        return false;
    }
}
