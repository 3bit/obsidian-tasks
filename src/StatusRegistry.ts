import { Status, StatusConfiguration } from './Status';

/**
 * Tracks all the registered statuses a task can have.
 *
 * @export
 * @class StatusRegistry
 */
export class StatusRegistry {
    private static instance: StatusRegistry;

    private _registeredStatuses: Status[] = [];

    /**
     * Creates an instance of Status and registers it for use. It will also check to see
     * if the default todo and done are registered and if not handle it internally.
     *
     * @memberof StatusRegistry
     */
    private constructor() {
        this.addDefaultStatusTypes();
    }

    /**
     * The static method that controls the access to the StatusRegistry instance.
     *
     * @static
     * @return {*}  {StatusRegistry}
     * @memberof StatusRegistry
     */
    public static getInstance(): StatusRegistry {
        if (!StatusRegistry.instance) {
            StatusRegistry.instance = new StatusRegistry();
        }

        return StatusRegistry.instance;
    }

    /**
     * Adds a new Status to the registry if not already registered.
     *
     * @param {StatusConfiguration | Status} status
     * @memberof StatusRegistry
     */
    public add(status: StatusConfiguration | Status): void {
        if (!this.hasIndicator(status.indicator)) {
            if (status instanceof Status) {
                this._registeredStatuses.push(status);
            } else {
                this._registeredStatuses.push(new Status(status));
            }
        }
    }

    /**
     * Returns the registered status by the indicator between the
     * square braces in the markdown task.
     *
     * @param {string} indicator
     * @return {*}  {Status}
     * @memberof StatusRegistry
     */
    public byIndicator(indicator: string): Status {
        if (this.hasIndicator(indicator)) {
            return this.getIndicator(indicator);
        }

        return Status.EMPTY;
    }

    /**
     * Resets the array os Status types to be empty.
     *
     * @memberof StatusRegistry
     */
    public clearStatuses(): void {
        this._registeredStatuses = [];
        this.addDefaultStatusTypes();
    }

    /**
     * To allow custom progression of task status each status knows
     * which status can come after it as a state transition.
     *
     * @return {*}  {Status}
     * @memberof StatusRegistry
     */
    public getNextStatus(status: Status): Status {
        if (status.nextStatusIndicator !== '') {
            const nextStatus = this.byIndicator(status.nextStatusIndicator);
            if (nextStatus !== null) {
                return nextStatus;
            }
        }
        return Status.EMPTY;
    }

    /**
     * Filters the Status types by the indicator and returns the first one found.
     *
     * @private
     * @param {string} indicatorToFind
     * @return {*}  {Status}
     * @memberof StatusRegistry
     */
    private getIndicator(indicatorToFind: string): Status {
        return this._registeredStatuses.filter(({ indicator }) => indicator === indicatorToFind)[0];
    }

    /**
     * Filters all the Status types by the indicator and returns true if found.
     *
     * @private
     * @param {string} indicatorToFind
     * @return {*}  {boolean}
     * @memberof StatusRegistry
     */
    private hasIndicator(indicatorToFind: string): boolean {
        return (
            this._registeredStatuses.find((element) => {
                return element.indicator === indicatorToFind;
            }) !== undefined
        );
    }

    /**
     * Checks the registry and adds the default status types.
     *
     * @private
     * @memberof StatusRegistry
     */
    private addDefaultStatusTypes(): void {
        const defaultStatuses = [Status.TODO, Status.DONE, Status.EMPTY];

        defaultStatuses.forEach((status) => {
            this.add(status);
        });
    }
}
