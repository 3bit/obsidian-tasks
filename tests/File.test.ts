import { readFileSync } from 'fs';
import { verify } from 'approvals/lib/Providers/Jest/JestApprovals';
import { findLineNumberOfTaskToToggle } from '../src/File';
import type { MockTogglingDataForTesting } from '../src/lib/MockDataCreator';

function testFindLineNumberOfTaskToToggle(
    jsonFileName: string,
    taskLineToToggle: string,
    expectedLineNumber: number | undefined,
    actualIncorrectLineFound?: string,
) {
    // Arrange
    const data = readFileSync('tests/__test_data__/MockDataForTogglingTasks/' + jsonFileName, 'utf-8');
    const everything: MockTogglingDataForTesting = JSON.parse(data);
    expect(everything.taskData.originalMarkdown).toEqual(taskLineToToggle);

    // Act
    const originalTask = everything.taskData;
    const fileLines = everything.fileData.fileLines;
    const listItemsCache = everything.cacheData.listItemsCache;

    let errorString: string | undefined;
    const captureError = (message: string) => {
        errorString = message;
    };

    const result = findLineNumberOfTaskToToggle(
        originalTask,
        fileLines,
        listItemsCache,
        captureError.bind(errorString),
    );

    // Assert
    verify(errorString);

    if (expectedLineNumber !== undefined) {
        expect(result).not.toBeUndefined();
        expect(result).toEqual(expectedLineNumber);

        const expectedLine = actualIncorrectLineFound ? actualIncorrectLineFound : everything.taskData.originalMarkdown;
        expect(everything.fileData.fileLines[result!]).toEqual(expectedLine);
    } else {
        expect(result).toBeUndefined();
    }
}

describe('replaceTaskWithTasks', () => {
    // --------------------------------------------------------------------------------
    // Issue 688
    describe('issue 688 - block referenced task', () => {
        const jsonFileName = '688_toggle_block_referenced_line_overwrites_wrong_line.json';
        const taskLineToToggle = '- [ ] #task task2b ^ca47c7';

        it.failing('correct behaviour', () => {
            // An incorrect line is currently found, so this test fails, due to bug 688
            const expectedLineNumber = 8;
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });

        it('current incorrect behaviour', () => {
            // An incorrect line is currently found, due to bug 688.
            // It is recognised as an incorrect line, and so line number is returned as undefined.
            const expectedLineNumber = undefined;
            // const actualIncorrectLineFound = '- [ ] #task task1a';
            testFindLineNumberOfTaskToToggle(jsonFileName, taskLineToToggle, expectedLineNumber);
        });
    });

    // --------------------------------------------------------------------------------
});
