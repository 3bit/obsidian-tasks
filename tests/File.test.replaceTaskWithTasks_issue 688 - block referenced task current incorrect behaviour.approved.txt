Tasks: Unable to find task in file Manual Testing/Task Toggling Scenarios/Embed Task in to Note.md.
Expected task:
- [ ] #task task2b ^ca47c7
Found task:
- [ ] #task task1a
